# Data Cleaning with Pandas

In [90]:
import pandas as pd

## Scenario

As data scientists, we want to build a model to predict the sale price of a house in Seattle in 2019, based on its square footage. We know that the King County Department of Assessments has comprehensive data available on real property sales in the Seattle area. We need to prepare the data.

### First, get the data!

When working on a project involving data that can fit on our computer, we store it in a `data` directory.

```bash
cd <project_directory>  # example: cd ~/flatiron_ds/pandas-3
mkdir data
cd data
```

Note that `<project_directory>` in angle brackets is a _placeholder_. You should type the path to the actual location on your computer where you're working on this project. Do not literally type `<project_directory>` and _do not type the angle brackets_. You can see an example in the _comment_ to the right of the command above.

Now, we'll need to download the two data files that we need. We can do this at the command line:

```bash
wget https://aqua.kingcounty.gov/extranet/assessor/Real%20Property%20Sales.zip
wget https://aqua.kingcounty.gov/extranet/assessor/Residential%20Building.zip
```

*Note:* If you do not have the `wget` command yet, you can install it: `brew install wget`.

Note that `%20` in a URL translates into a space. Even though you will *never put spaces in filenames*, you may need to deal with spaces that _other_ people have used in filenames.

There are two ways to handle the spaces in these filenames when referencing them at the command line.

#### 1. You can _escape_ the spaces by putting a backslash (`\`, remember _backslash is next to backspace_) before each one:

`unzip Real\ Property\ Sales.zip`

This is what happens if you tab-complete the filename in the terminal. Tab completion is your friend!

#### 2. You can put the entire filename in quotes:

`unzip "Real Property Sales.zip"`

Try unzipping these files with the `unzip` command. The `unzip` command takes one argument, the name of the tile that you want to unzip.


You can use tab completion (press the `tab` key after the first three letters) to fill in the names, including spaces. This will 

In [190]:
sales_df = pd.read_csv('data/EXTR_RPSale.csv')

/anaconda3/envs/learn-env/lib/python3.7/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (1,2) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Seeing pink? Warnings are useful!

Note the warning above: `DtypeWarning: Columns (1, 2) have mixed types.` Because we start with an index of zero, the columns that we're being warned about are actually the _second_ and _third_ columns, `sales_df['Major']` and `sales_df['Minor']`.

In [88]:
sales_df.head().T

0  \
ExciseTaxNbr                                                  2687551   
Major                                                          138860   
Minor                                                             110   
DocumentDate                                               08/21/2014   
SalePrice                                                      245000   
RecordingNbr                                           20140828001436   
Volume                                                                  
Page                                                                    
PlatNbr                                                                 
PlatType                                                                
PlatLot                                                                 
PlatBlock                                                               
SellerName          WENKLE NOEL SMITH -TRUSTEE                    ...   
BuyerName           ALEXANDER APRIL                               ...   
PropertyType                                                        3   
PrincipalUse                                                        6   
SaleInstrument                                                      3   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                          1   
PropertyClass                                                       8   
SaleWarning                                                             

                                                                    1  \
ExciseTaxNbr                                                  1235111   
Major                                                          664885   
Minor                                                              40   
DocumentDate                                               07/09/1991   
SalePrice                                                           0   
RecordingNbr                                           199203161090     
Volume                                                            071   
Page                                                              001   
PlatNbr                                                        664885   
PlatType                                                            C   
PlatLot                                                B102             
PlatBlock                                                               
SellerName          HERRMAN JEFFREY A+PAULINE A                   ...   
BuyerName           HERRMAN PAULINE A TTEE                        ...   
PropertyType                                                        3   
PrincipalUse                                                        0   
SaleInstrument                                                     26   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                         18   
PropertyClass                                                       3   
SaleWarning                                                        11   

                                                                    2  \
ExciseTaxNbr                                                  2704079   
Major                                                          423943   
Minor                                                              50   
DocumentDate                                               10/11/2014   
SalePrice                                          

### Data overload?

That's a lot of columns. We're only interested in identifying the date, sale price, and square footage of each specific property. What can we do?

In [191]:
sales_df = sales_df[['Major', 'Minor', 'DocumentDate', 'SalePrice']]

In [192]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2031737 entries, 0 to 2031736
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 62.0+ MB


In [193]:
bldg_df = pd.read_csv('data/Residential Building.zip')

/anaconda3/envs/learn-env/lib/python3.7/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Another warning! Which column has index 11?

In [93]:
bldg_df.columns[11]

IndexError: index 11 is out of bounds for axis 0 with size 4

`ZipCode` seems like a potentially useful column. We'll need it to determine which house sales took place in Seattle.

In [194]:
bldg_df.head().T

0                          1  \
Major                                   14800                      14800   
Minor                                     168                        425   
BldgNbr                                     1                          1   
NbrLivingUnits                              1                          1   
Address             3047   63RD AVE SW  98116  3231   63RD AVE SW  98116   
BuildingNumber                          3047                       3231    
Fraction                                                                   
DirectionPrefix                                                            
StreetName          63RD                       63RD                        
StreetType                             AVE                        AVE      
DirectionSuffix                            SW                         SW   
ZipCode                                 98116                      98116   
Stories                                     1                          2   
BldgGrade                                   7                          8   
BldgGradeVar                                0                          0   
SqFt1stFloor                              750                       1010   
SqFtHalfFloor                               0                          0   
SqFt2ndFloor                                0                        560   
SqFtUpperFloor                              0                          0   
SqFtUnfinFull                               0                          0   
SqFtUnfinHalf                               0                          0   
SqFtTotLiving                            1500                       2070   
SqFtTotBasement                           750                        680   
SqFtFinBasement                           750                        500   
FinBasementGrade                            7                          7   
SqFtGarageBasement                          0                          0   
SqFtGarageAttached                          0                          0   
DaylightBasement                            N                              
SqFtOpenPorch                              20                         60   
SqFtEnclosedPorch                           0                          0   
SqFtDeck                                    0                        290   
HeatSystem                                  5                          5   
HeatSource                                  2                          2   
BrickStone                                  0                          0   
ViewUtilization                             N                              
Bedrooms                                    3                          4   
BathHalfCount                               0                          1   
Bath3qtrCount                               0                          1   
BathFullCount                               2                          1   
FpSingleStory                               0                          0   
FpMultiStory                                0                          1   
FpFreestanding                              0                          1   
FpAdditional                                0                          0   
YrBuilt                                  1949                       1924   
YrRenovated                                 0                          0   
PcntComplete                                0                          0   
Obsolescence                                0                          0   
PcntNetCondition                            0                          0   
Condition                                   4                          5   
AddnlCost                                3000                          0   

                                            2                          3  \
Major                                   14800                      14800   
Minor          

### So many features!

As data scientists, we should be _very_ cautious about discarding potentially useful data. But, today, we're interested in _only_ the total square footage of each property. What can we do?


In [195]:
bldg_df = bldg_df[['Major', 'Minor', 'SqFtTotLiving', 'ZipCode']]

In [196]:
bldg_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 512387 entries, 0 to 512386
Data columns (total 4 columns):
Major            512387 non-null int64
Minor            512387 non-null int64
SqFtTotLiving    512387 non-null int64
ZipCode          468443 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB


In [166]:
sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])

ValueError: You are trying to merge on object and int64 columns. If you wish to proceed you should use pd.concat

### Error!

Why are we seeing an error when we try to join the dataframes?

<table>
    <tr>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2013160 entries, 0 to 2013159
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.4+ MB</pre></td>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511359 entries, 0 to 511358
Data columns (total 4 columns):
Major            511359 non-null int64
Minor            511359 non-null int64
SqFtTotLiving    511359 non-null int64
ZipCode          468345 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB
</pre></td>
    </tr>
</table>

Review the error message in light of the above:

* `ValueError: You are trying to merge on object and int64 columns.`

In [ ]:
sales_df['Major'] = pd.to_numeric(sales_df['Major'])

### Error!

Note the useful error message above:

`ValueError: Unable to parse string "      " at position 936643`

In this case, we want to treat non-numeric values as missing values. Let's see if there's a way to change how the `pd.to_numeric` function handles errors.

In [ ]:
# The single question mark means "show me the docstring"
pd.to_numeric?

Here's the part that we're looking for:
```
errors : {'ignore', 'raise', 'coerce'}, default 'raise'
    - If 'raise', then invalid parsing will raise an exception
    - If 'coerce', then invalid parsing will be set as NaN
    - If 'ignore', then invalid parsing will return the input
```

Let's try setting the `errors` parameter to `'coerce'`.

In [197]:
sales_df['Major'] = pd.to_numeric(sales_df['Major'], errors='coerce')

Did it work?

In [198]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2031737 entries, 0 to 2031736
Data columns (total 4 columns):
Major           float64
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: float64(1), int64(1), object(2)
memory usage: 62.0+ MB


It worked! Let's do the same thing with the `Minor` parcel number.

In [199]:
sales_df['Minor'] = pd.to_numeric(sales_df['Minor'], errors='coerce')

In [200]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2031737 entries, 0 to 2031736
Data columns (total 4 columns):
Major           float64
Minor           float64
DocumentDate    object
SalePrice       int64
dtypes: float64(2), int64(1), object(1)
memory usage: 62.0+ MB


Now, let's try our join again.

In [201]:
sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])

In [202]:
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

In [203]:
import numpy as np
non_zero_mask = (sales_data['SalePrice']> 0)
np.percentile(sales_data['SalePrice'][non_zero_mask], 4)

65000.0

In [168]:
sales_mini = sales_data.iloc[-1000:].copy()

In [169]:
sales_mini['Date'] = sales_mini['DocumentDate'].apply(pd.Timestamp)

In [170]:
sales_mini.head().T

1448987              1448988              1448990  \
Major                       713780               713780               446590   
Minor                          330                  330                   50   
DocumentDate            05/21/1997           08/12/2003           05/22/1995   
SalePrice                   128000               197000               124000   
SqFtTotLiving                 1660                 1660                 1200   
ZipCode                      98023                98023                98133   
Date           1997-05-21 00:00:00  2003-08-12 00:00:00  1995-05-22 00:00:00   

                           1448991              1448992  
Major                        25980               152407  
Minor                          390                 9151  
DocumentDate            08/06/1990           05/18/2017  
SalePrice                   162000               612000  
SqFtTotLiving                 1250                 1980  
ZipCode                      98008                98024  
Date           1990-08-06 00:00:00  2017-05-18 00:00:00

In [105]:
ts = pd.Timestamp('05/13/1985')

In [106]:
ts.year

1985

In [107]:
sales_mini['Year'] = sales_mini.apply(lambda ts: ts.year))

SyntaxError: invalid syntax (<ipython-input-107-663b5ca0b49b>, line 1)

In [171]:
sales_mini.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1000 entries, 1448987 to 1450454
Data columns (total 7 columns):
Major            1000 non-null float64
Minor            1000 non-null float64
DocumentDate     1000 non-null object
SalePrice        1000 non-null int64
SqFtTotLiving    1000 non-null int64
ZipCode          1000 non-null object
Date             1000 non-null datetime64[ns]
dtypes: datetime64[ns](1), float64(2), int64(2), object(2)
memory usage: 62.5+ KB


In [204]:
sales_data['Date'] = sales_data['DocumentDate'].apply(pd.Timestamp)

In [205]:
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode       Date
0  138860.0  110.0   08/21/2014     245000           1490   98002 2014-08-21
1  138860.0  110.0   06/12/1989     109300           1490   98002 1989-06-12
2  138860.0  110.0   01/16/2005      14684           1490   98002 2005-01-16
3  138860.0  110.0   06/08/2005          0           1490   98002 2005-06-08
4  423943.0   50.0   10/11/2014          0            960   98092 2014-10-11

In [206]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1450456 entries, 0 to 1450455
Data columns (total 7 columns):
Major            1450456 non-null float64
Minor            1450456 non-null float64
DocumentDate     1450456 non-null object
SalePrice        1450456 non-null int64
SqFtTotLiving    1450456 non-null int64
ZipCode          1332401 non-null object
Date             1450456 non-null datetime64[ns]
dtypes: datetime64[ns](1), float64(2), int64(2), object(2)
memory usage: 88.5+ MB


We can see right away that we're missing zip codes for many of the sales transactions. (1321536 non-null entries for ZipCode is fewer than the 1436772 entries in the dataframe.) 

In [207]:
sales_data.loc[sales_data['ZipCode'].isna()].head(10)

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
91   717370.0   350.0   12/01/1997          0           3380     NaN   
92   717370.0   350.0   09/13/2004     300000           3380     NaN   
93   717370.0   350.0   02/06/2006     901000           3380     NaN   
110  277110.0  1923.0   02/08/2007     372500           1000     NaN   
111  277110.0  1923.0   02/08/2007          0           1000     NaN   
236  720319.0   520.0   11/20/2013     699950           2840     NaN   
237  720319.0   520.0   09/16/2013          0           2840     NaN   
247  894677.0   240.0   12/21/2016     818161           2450     NaN   
273  935840.0   290.0   06/10/2005    2025000           3144     NaN   
274  935840.0   290.0   06/20/2019          0           3144     NaN   

          Date  
91  1997-12-01  
92  2004-09-13  
93  2006-02-06  
110 2007-02-08  
111 2007-02-08  
236 2013-11-20  
237 2013-09-16  
247 2016-12-21  
273 2005-06-10  
274 2019-06-20

Because we are interested in finding houses in Seattle zip codes, we will need to drop the rows with missing zip codes.

In [208]:
sales_data = sales_data.loc[~sales_data['ZipCode'].isna(), :] # ~ refers not in pandas/numpy

sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode       Date
0  138860.0  110.0   08/21/2014     245000           1490   98002 2014-08-21
1  138860.0  110.0   06/12/1989     109300           1490   98002 1989-06-12
2  138860.0  110.0   01/16/2005      14684           1490   98002 2005-01-16
3  138860.0  110.0   06/08/2005          0           1490   98002 2005-06-08
4  423943.0   50.0   10/11/2014          0            960   98092 2014-10-11

# Your turn: Data Cleaning with Pandas

### 1. Investigate and drop rows with invalid values in the SalePrice and SqFtTotLiving columns.

Use multiple notebook cells to accomplish this! Press `[esc]` then `B` to create a new cell below the current cell. Press `[return]` to start typing in the new cell.

In [227]:
sales_data.loc[sales_data['SalePrice'] <= 20000].head(10)

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
2     138860.0   110.0   01/16/2005      14684           1490   98002   
90    211430.0   160.0   01/03/1994      12250           1840   98106   
218   563450.0   966.0   10/15/1999          1           1880   98028   
240   154580.0  1220.0   01/10/2018         10           1320   98038   
282   935840.0   280.0   10/03/2006       1000           3224   98003   
707   200900.0   720.0   06/08/1993      18000           2900   98198   
1224  914110.0    15.0   10/05/1985      20000           2500   98133   
1520  607130.0   720.0   06/18/2003       3000           2520   98006   
1600  438920.0   728.0   02/19/2005        400           1880   98004   
1604  438920.0   728.0   03/01/2005        400           1880   98004   

           Date  
2    2005-01-16  
90   1994-01-03  
218  1999-10-15  
240  2018-01-10  
282  2006-10-03  
707  1993-06-08  
1224 1985-10-05  
1520 2003-06-18  
1600 2005-02-19  
1604 2005-03-01

In [210]:
sales_data.loc[sales_data['SqFtTotLiving'] <= 600].head(10)

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
1798  382220.0   41.0   04/25/2012     475000            220   98125   
1800  382220.0   41.0   04/25/2012     475000            370   98125   
1801  382220.0   41.0   04/25/2012     475000            290   98125   
1802  382220.0   41.0   04/25/2012     475000            460   98125   
1803  382220.0   41.0   02/02/1995          0            220   98125   
1805  382220.0   41.0   02/02/1995          0            370   98125   
1806  382220.0   41.0   02/02/1995          0            290   98125   
1807  382220.0   41.0   02/02/1995          0            460   98125   
1808  382220.0   41.0   11/14/2008          0            220   98125   
1810  382220.0   41.0   11/14/2008          0            370   98125   

           Date  
1798 2012-04-25  
1800 2012-04-25  
1801 2012-04-25  
1802 2012-04-25  
1803 1995-02-02  
1805 1995-02-02  
1806 1995-02-02  
1807 1995-02-02  
1808 2008-11-14  
1810 2008-11-14

In [211]:
sales_data = sales_data.loc[sales_data['SalePrice'] != 0, :] # ~ refers not in pandas/numpy
sales_data = sales_data.loc[sales_data['SqFtTotLiving'] != 0, :] 
sales_data = sales_data.loc[sales_data['SalePrice'] > 0, :] # ~ refers not in pandas/numpy
sales_data = sales_data.loc[sales_data['SqFtTotLiving'] > 0, :] 
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode       Date
0  138860.0  110.0   08/21/2014     245000           1490   98002 2014-08-21
1  138860.0  110.0   06/12/1989     109300           1490   98002 1989-06-12
2  138860.0  110.0   01/16/2005      14684           1490   98002 2005-01-16
6  423943.0   50.0   07/15/1999      96000            960   98092 1999-07-15
7  423943.0   50.0   01/08/2001     127500            960   98092 2001-01-08

### 2. Investigate and handle non-numeric ZipCode values

Can you find a way to shorten ZIP+4 codes to the first five digits?

What's the right thing to do with missing values?

In [289]:
sales_data_copy = sales_data

In [290]:
# Read the error message and decide how to fix it.
# Note: using errors='coerce' is the *wrong* choice in this case.
def is_integer(x):
    try:
        _ = int(x)
    except ValueError:
        return False
    return True

stringed_zips = sales_data_copy.loc[sales_data_copy['ZipCode'].apply(is_integer) == False, 'ZipCode'] 
# sales_data.loc[sales_data['ZipCode'].apply(is_integer) == False, 'ZipCode'].head()
stringed_zips.head()

17669    98042-3001
17670    98042-3001
17671    98042-3001
17672    98042-3001
21350    98199-3014
Name: ZipCode, dtype: object

In [228]:
# df['text'].map(lambda review_text: len(review_text.split())).head()
# sales_data_copy['ZipCode'].map(lambda i: str('sales_data_copy['ZipCode']')).head()


In [291]:
# cleaned_zips = stringed_zips.aookt(lambda)
def zipcode_short(x):
    return str(x)[:5]

In [292]:
sales_data_copy['New_zip'] = sales_data_copy['ZipCode'].map(zipcode_short)
sales_data_copy.loc[sales_data_copy['New_zip'].apply(is_integer) == False, 'New_zip'].head()

329722     
329723     
329724     
398956    B
436072    A
Name: New_zip, dtype: object

In [293]:
sales_data_copy.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode       Date  \
0  138860.0  110.0   08/21/2014     245000           1490   98002 2014-08-21   
1  138860.0  110.0   06/12/1989     109300           1490   98002 1989-06-12   
2  138860.0  110.0   01/16/2005      14684           1490   98002 2005-01-16   
6  423943.0   50.0   07/15/1999      96000            960   98092 1999-07-15   
7  423943.0   50.0   01/08/2001     127500            960   98092 2001-01-08   

  New_zip  
0   98002  
1   98002  
2   98002  
6   98092  
7   98092

### 3. Add a column for PricePerSqFt



In [295]:
sales_data_copy['PricePerSqFt'] = sales_data_copy['SalePrice'] / sales_data_copy['SqFtTotLiving']

In [296]:
sales_data_copy.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode       Date  \
0  138860.0  110.0   08/21/2014     245000           1490   98002 2014-08-21   
1  138860.0  110.0   06/12/1989     109300           1490   98002 1989-06-12   
2  138860.0  110.0   01/16/2005      14684           1490   98002 2005-01-16   
6  423943.0   50.0   07/15/1999      96000            960   98092 1999-07-15   
7  423943.0   50.0   01/08/2001     127500            960   98092 2001-01-08   

  New_zip  PricePerSqFt  
0   98002    164.429530  
1   98002     73.355705  
2   98002      9.855034  
6   98092    100.000000  
7   98092    132.812500

### 4. Subset the data to 2019 sales only.

We can assume that the DocumentDate is approximately the sale date.

In [306]:
# df.date.map(lambda x: x[:4]).head()
sales_data_copy['Date_Year'] = sales_data_copy.DocumentDate.map(lambda x: x[6:11])
sales_data_copy.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode       Date  \
0  138860.0  110.0   08/21/2014     245000           1490   98002 2014-08-21   
1  138860.0  110.0   06/12/1989     109300           1490   98002 1989-06-12   
2  138860.0  110.0   01/16/2005      14684           1490   98002 2005-01-16   
6  423943.0   50.0   07/15/1999      96000            960   98092 1999-07-15   
7  423943.0   50.0   01/08/2001     127500            960   98092 2001-01-08   

  New_zip  PricePerSqFt Date_Year  
0   98002    164.429530      2014  
1   98002     73.355705      1989  
2   98002      9.855034      2005  
6   98092    100.000000      1999  
7   98092    132.812500      2001

In [309]:
sales_data_copy.loc[sales_data_copy['Date_Year'] == "2019"].head()

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
396  302605.0  9246.0   05/07/2019    1010000           2260   98034   
477   30200.0   645.0   03/04/2019     380000           1750   98001   
800   42106.0  9050.0   06/01/2019     755000           2170   98010   
816  214090.0   110.0   03/06/2019     390000           1780   98042   
966  145360.0  2285.0   05/02/2019     690000           1910   98125   

          Date New_zip  PricePerSqFt Date_Year  
396 2019-05-07   98034    446.902655      2019  
477 2019-03-04   98001    217.142857      2019  
800 2019-06-01   98010    347.926267      2019  
816 2019-03-06   98042    219.101124      2019  
966 2019-05-02   98125    361.256545      2019

### 5. Subset the data to zip codes within the City of Seattle.

You'll need to find a list of Seattle zip codes!

In [311]:
Seattle_zip = [98101, 98102, 98103, 98104, 98105, 98106, 98107, 98108, 98109, 98112, 
               98115, 98116, 98117, 98118, 98119, 98121, 98122, 98125, 98126, 98133, 
               98134, 98136, 98144, 98146, 98154, 98164, 98174, 98177, 98178, 98195, 98199]

In [315]:

sales_data_copy[sales_data_copy['ZipCode'].isin(Seattle_zip)].head()

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
146  510140.0  4256.0   08/01/2003     235000            870   98115   
147  510140.0  4256.0   03/20/1996     135000            870   98115   
149  510140.0  4256.0   07/09/2013     275000            870   98115   
453  199420.0   405.0   01/22/1993     172000           1330   98103   
464  982920.0  1000.0   08/28/2018     575000           1210   98122   

          Date New_zip  PricePerSqFt Date_Year  
146 2003-08-01   98115    270.114943      2003  
147 1996-03-20   98115    155.172414      1996  
149 2013-07-09   98115    316.091954      2013  
453 1993-01-22   98103    129.323308      1993  
464 2018-08-28   98122    475.206612      2018

### 6. What is the mean price per square foot for a house sold in Seattle in 2019?

Don't just type the answer. Type code that generates the answer as output!